iucvterm/doc/iucvconn.1
.\" iucvconn.1
.\"
.\"
.\" Copyright IBM Corp. 2008, 2017
.\" s390-tools is free software; you can redistribute it and/or modify
.\" it under the terms of the MIT license. See LICENSE for details.
.\" ----------------------------------------------------------------------
.TH iucvconn "1" "March 2009" "s390-tools" "IUCV terminal applications"
.
.ds i \fBiucvconn\fP
.
.
.
.SH NAME
iucvconn \- start terminal connection over z/VM IUCV
.
.
.
.SH SYNOPSIS
.B iucvconn
.RB [ \-h | \-\-help ]
.br
.B iucvconn
.RB [ \-v | \-\-version ]
.br
.B iucvconn
.RB [ \-e | \-\-escape-char
.IR escape_char ]
.RB [ \-s | \-\-sessionlog
.IR log_file ]
.I vm_guest
.I terminal_id
.
.
.
.SH DESCRIPTION
The \*i application establishes a terminal connection to another
Linux instance. The Linux instance where \*i runs and the target Linux instance
must be z/VM guest operating systems of the same z/VM instance.
The communication path between both virtual machines is based on the z/VM
Inter-User Communication Vehicle (IUCV).
Because z/VM IUCV is independent from TCP/IP, Linux instances with no external
network can be accessed.
.PP
\*i accesses the terminal specified by \fIterminal_id\fP on the z/VM guest
virtual machine \fIvm_guest\fP.
The \fIterminal_id\fP is similar to a port number in TCP/IP communications.
.
\fIterminal_id\fP is case-sensitive and consists of up to 8 alphanumeric
characters. \fIterminal_id\fP must either identify a running \fBiucvtty\fP
instance, or a terminal device provided by the z/VM IUCV hypervisor console
(HVC) device driver.
.
.
.
.SH OPTIONS
.TP
.BR \-h ", " \-\-help
Display a short help text, then exit.
.
.TP
.BR \-v ", " \-\-version
Display the version number of \*i, then exit.
.
.TP
.BR \-e ", " \-\-escape-char\~\fIescape_char\fP
Set an escape character for this terminal session. You need an escape character
to access special \*i functions. The default escape character is underscore (_).
If \fIescape_char\fP is set to "none", escaping is not possible.
The escape character can be the closing bracket (]), the caret (^), the
underscore (_), or any alphabetical character except C, D, Q, S and Z.
The \fIescape_char\fP character is not case-sensitive.

To invoke a \*i function, press Ctrl\^+\^\fIescape_char\fP, followed
by the function character. For example, press "Ctrl\^+\^_ d" to disconnect the
terminal session and exit \*i.

For details, see section "ESCAPE CHARACTERS" below.
.
.TP
.BR \-s ", " \-\-sessionlog\~\fIlogfile\fP
Create a transcript of the terminal session and write session data to three
different files.

\fIlog_file\fP contains the raw terminal data stream. The
.I log_file\fB.timing\fP
file contains timing data that can be used for replaying the raw terminal data
stream using realistic output delays.
Additional terminal session information is logged to
.IR log_file\fB.info\fP "."

If any of these files already exist, \*i exits with an error. To proceed
either delete the files or choose another file name for \fIlog_file\fP.
.
.
.
.SH "ESCAPE CHARACTERS"
\*i supports functions through the use of an escape character.
The default escape character (\fIescape_char\fP) is underscore (_).

To enter the "escape mode" of \*i, press the Control and the \fIescape_char\fP
key together ("Ctrl\^+\^_") In the "escape mode", \*i expects a single character
designating the special function.

The special functions are (assuming \fIescape_char\fP is set to underscore (_)):
.RS 4
.IP "\fBCtrl\^+\^_\fP" 8
Send the escape character to the connected terminal.
.IP "\fB.\fP or \fBd\fP" 8
Disconnect and exit \*i.
.IP "\fBr\fP" 8
Force resizing of the connected terminal.
.RE
.PP
Any other single character is ignored and is not send to the terminal. If
multiple characters are entered together, \*i leaves the "escape mode" and sends
the characters to the connected terminal.
.
.\" FIXME: change "connected terminal" to terminal
.
.
.SH "EXAMPLES"
To access the "lnxterm" terminal on the Linux instance in
z/VM guest virtual machine LNXSYS01:
.PP
.ft CR
.in +0.25in
.nf
iucvconn LNXSYS01 lnxterm
.fi
.in -0.25in
.ft
.PP
To access the first z/VM IUCV HVC terminal on the Linux instance in
z/VM guest virtual machine LNXSYS02:
.PP
.ft CR
.in +0.25in
.nf
iucvconn LNXSYS02 lnxhvc0
.fi
.in -0.25in
.ft
.PP
To create a transcript of the terminal session to the Linux instance in
z/VM guest virtual machine LNXSYS99:
.PP
.ft CR
.in +0.25in
.nf
iucvconn -s ~/transcripts/lnxsys99 LNXSYS99 lnxhvc0
.fi
.in -0.25in
.ft
.
.
.
.SH ENVIRONMENT
.TP
.B TERM
The terminal environment variable contains the name of the terminal
(e.g. linux, xterm, ...) and also specifies the terminal capabilities.

\*i transfers the value of the \fBTERM\fP environment variable at connection
startup to the target system. The \fBiucvtty\fP application sets the
\fBTERM\fP environment variable on the target system before the login program
is started.
Setting the \fBTERM\fP environment variable ensures that the terminal
capabilities are the same on both sides of the connection.

The z/VM IUCV hypervisor console (HVC) device driver ignores the content of
the \fBTERM\fP environment variable transferred by \*i. See the
.BR hvc_iucv (9)
manual page for instructions to configure the terminal environment.
.
.
.
.SH DIAGNOSTICS
If the Linux kernel does not include kernel support for the AF_IUCV network
addressing family, \*i exits and displays the message
.I 'AF_IUCV address family is not available: Address family not supported by
.IR protocol' "."
.
.
.
.SH SECURITY
.SS Linux
The \*i program does not require superuser privileges and can be used by
regular users.  Each connection attempt and failure is logged to the
.BR authpriv
syslog facility.
.\".PP
.\"Depending on the login program started by the \fBiucvtty\fP program,
.\"the user must authenticate to the remote system.
.PP
Terminal session transcripts facilitate logging and auditing of terminal
activity on remote systems or recording and playing back terminal sessions,
for example for education and presentations.
.
.
.SS z/VM guest virtual machine
The z/VM guest virtual machine must have authorization to establish IUCV
communication paths to other z/VM guest virtual machines.

See the
.BR af_iucv (7)
manual page for details about IUCV authorizations.
.
.
.
.SH NOTES
The Linux instance where \*i runs and the target Linux instances
must be z/VM guest operating systems of the same z/VM instance.
.
.
.
.SH "SEE ALSO"
.BR iucvtty (1),
.BR hvc_iucv (9),
.BR af_iucv (7),
.BR scriptreplay (1)

.I "Linux on System z - Device Drivers, Features, and Commands"


CHANGELOG.md
Release history for s390-tools (MIT version)
--------------------------------------------

* __v2.xx.x (20xx-xx-xx)__

  For Linux kernel version: 6.x

  Add new tools / libraries:

  Changes of existing tools:

  Bug Fixes:

* __v2.38.0 (2025-06-25)__

  For Linux kernel version: 6.14 / 6.15

  Add new tools:
  - udev: New rule to set newly hotplugged CPUs online
  - zmemtopo: Display memory topology information
  - zpwr: Display power readings of a partition and CPC

  Removed tools / features:
  - check_hostkeydoc: Remove installation target
  - scsi_logging_level: Delete SCSI logging script (available in sg3_utils)
  - zdump: Drop build_arch for s390 DASD dumps
  - zdump: Drop non-extended multi-volume DASD dump support
  - zdump: Drop support of 32-bit dump architecture
  - zdump: Drop support of non-extended single volume DASD dumpers
  - zdump: Drop support of obsolete dumps and dumpers

  Changes of existing tools / libraries:
  - Various man-pages fixes
  - check_hostkeydoc: Add deprecation warning
  - check_hostkeydoc: Move to scripts directory
  - cpuplugd: Allow cpu hotplugging on systems without polarization
  - dbginfo.sh: Add Ubuntu snap tool
  - dbginfo.sh: Add missing config data and logs
  - dbginfo.sh: Reworking the container section
  - dbginfo.sh: Update for network commands
  - dbginfo.sh: Updating info for disks and lvm
  - libutil: Add machine type definition for machines 9175 and 9176
  - lscpumf: Add support for IBM z17 counter sets
  - lshwc: Add command line flag for run time
  - lshwc: Add flags to display counter values in hex
  - lshwc: Add output '--format' option
  - lshwc: Add support for delta counter value display
  - lspai: Add output '--format' option
  - lsreipl: Add secure boot state to output
  - lswhc: Add short names to lshwc output
  - pv_tools: Add Bash and Zsh completions
  - pvapconfig: Add '--unbind' option
  - pvimg/boot: Print error messages from stage3a bootloader
  - pvimg: Add support for CCK update
  - pvsecret: Add support for CCK update
  - pvsecret: Allow retrieving secrets by index & warn for duplicated entries
  - pvsecret: Deny adding secrets with duplicated secret IDs
  - zdev: Add support for virtio devices
  - zipl: Enhance mirror support
  - zipl: Implement '--dry-run' option for all dump jobs
  - zipl_helper.device-mapper: Support mirrors over NVMe devices
  - zkey/dracut: Add a dracut config file for zkey
  - zkey/initramfs: Update initramfs hook to correct drivers and include zkey plugins
  - zkey: Add support for converting a clear-key LUKS2 volume to use a secure key

  Bug Fixes:
  - chpstat: Add missing CMG 5 data fields
  - chpstat: Fix DPU utilization calculation
  - libutil/util_file: Handle over-read in util_file_read_fd()
  - pvattest: Fix successful 'check' evaluation
  - pvsecret: Fix some edge cases for plaintext keys
  - zipl_helper.device-mapper: Fix imprecise is_device_mapper() predicate
  - zkey: Fix EP11 secure key reencipher function
  - zpcictl: Fix command line parsing for invalid options

* __v2.37.0 (2025-02-07)__

  For Linux kernel version: 6.13

  Changes of existing tools:
  - dbginfo.sh: Add details on CPU-measurement
  - dbginfo.sh: Add new crypto command
  - dbginfo.sh: Add overview commands and crypto update
  - dbginfo.sh: Adding kdump info
  - dbginfo.sh: Removing outdated email references
  - dbginfo.sh: Rework network section
  - dbginfo.sh: Update copyright 2nd year
  - pvimg: Add '--(enable|disable)-image-encryption' flags to 'pvimg create'
  - pvimg: Add '--cck <FILE>' command line option and make '--comm-key' an alias
  - pvimg: Add '--hdr-key' command line option to 'pvimg create'
  - pvimg: Rename '--key' into '--hdr-key' and use '--key' as an alias (for 'pvimg info')
  - pvsecret: Add support for retrievable secrets
  - ziorep_config: Add PCHID field to adapter report
  - ziorep_traffic: Add DEVBUSID column to traffic report
  - ziorep_utilization: Add --fcp-device parameter to print virtual adapter report
  - ziorep_utilization: Add PCHID column to physical adapter report
  - ziorep_utilization: Now prints only physical adapter report by default
  - ziorep_utilization: Swap Bus-ID and CHPID columns in virtual adapter report
  - zipl/boot: Increase section size for eckd_mv dumper
  - zkey: Add support for listing and importing protected virtualization secrets

  Bug Fixes:
  - chpstat: Fix invalid utilization data on older kernels
  - opticsmon: Fix runaway loop in on_link_change()
  - zipl: Update inline assembly for GCC 15
  - zipl_helper.device-mapper: Add missed step in logical device resolution

* __v2.36.0 (2024-12-06)__

  For Linux kernel version: 6.12

  s390-tools: Define Rust MSRV as 1.75.0

  Add new tools / libraries:
  - cpacfinfo: Tool to provide CPACF information
  - opticsmon: Tools to monitor optical modules for directly attached PCI based NICs
  - pvimg: Rust rewrite of genprotimg

  Changes of existing tools:
  - chpstat: Add data bandwidth utilization column
  - chpstat: Add support for full CMCB
  - chpstat: Add support for new CMG types
  - dbginfo.sh: add overview commands and crypto update
  - genprotimg:
    1. genprotimg is now a symbolic link to the new tool `pvimg create`
    2. Breaking API changes in genprotimg:
       1. An existing output file is no longer silently overwritten; to revert
          to the old behavior use the `--overwrite` option.
       2. The Linux kernel component is now checked to determine whether it is a
          binary s390x kernel. To disable this behavior use the
          `--no-component-check` option.
  - hyptop: Support for structured output (json, json-seq, csv)
  - lszfcp: Add missing fallback marker for non-good fc_host port_state
  - lszfcp: Improve speed with many SCSI devices
  - pvattest: Add attestation policy check command
  - zipl: Add support of partitions of mirror md-devices

  Bug Fixes:
  - lszcrypt: Fix wrong state showing up for removed AP queue within SE guest

iucvterm/src/iucvconn.c
/*
 * iucvconn - Application that establishes a terminal connection over IUCV
 *
 * Core application
 *
 * Copyright IBM Corp. 2008, 2017
 *
 * s390-tools is free software; you can redistribute it and/or modify
 * it under the terms of the MIT license. See LICENSE for details.
 */
#include <errno.h>
#include <pwd.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/socket.h>
#include <sys/types.h>
#include <syslog.h>
#include <termios.h>
#include <unistd.h>

#include "lib/util_base.h"

#include "iucvterm/config.h"
#include "iucvterm/functions.h"
#include "iucvterm/gettext.h"

#define SYSLOG_IDENT		"iucvconn"
#define PRG_COMPONENT		SYSLOG_IDENT
#define DEFAULT_TERM		"linux"

#define AUDIT(f, ...)		do { \
					syslog(LOG_INFO, (f), __VA_ARGS__); \
					write_session_info((f), __VA_ARGS__); \
				} while (0);

/* escape mode actions */
enum esc_action_t {
	DISCONNECT,		/* Disconnect from terminal   */
	RESIZE,			/* Force terminal resizing    */
	SEND,			/* Send data (default action) */
	IGNORE,			/* Ignore escape character   */
};


static volatile sig_atomic_t resize_tty;
static struct termios ios_orig;		/* store original termio settings */


/**
 * sig_handler() - Signal handler
 * @sig:	Signal number
 */
static void sig_handler(int sig)
{
	switch (sig) {
	case SIGWINCH:
		resize_tty = sig;
		break;

	case SIGTERM:
		tcsetattr(STDIN_FILENO, TCSANOW, &ios_orig);
		close_session_log();
		_exit(0);
		break;
	}
}

/**
 * get_msg_char() - Returns single character from message
 * @msg:	The IUCV terminal message
 *
 * Returns the (single) character of an IUCV terminal message
 * if @msg is of type MSG_TYPE_DATA and contains a single character
 * (datalen == 1). Otherwise the routine returns zero.
 */
static unsigned char get_msg_char(const struct iucvtty_msg *msg)
{
	if (msg->type != MSG_TYPE_DATA || msg->datalen != 1)
		return 0;

	return msg->data[0];
}

/**
 * is_esc_char() - Check message for escape character
 * @msg:	The IUCV terminal message
 * @esc:	The escape character
 *
 * Returns 1 if @msg contains the escape character @esc only; otherwise the
 * function returns 0.
 */
static int is_esc_char(const struct iucvtty_msg *msg, unsigned char esc)
{
	if (!esc)
		return 0;

	return (get_msg_char(msg) == esc) ? 1 : 0;
}

/**
 * get_action() - Returns the action from an escaped character
 * @msg:	The IUCV terminal message
 * @esc:	The escape character
 *
 * Returns the appropriate action of the escaped character.
 * The action is derived from the single character stored in the IUCV terminal
 * message @msg. If the escape character is recognized, SEND is returned for
 * sending the "escaped" escape character to the terminal.
 *
 * If it contains multiple characters, the default SEND action is used to
 * indicate that the escape mode is done and to force sending the complete data
 * characters (e.g. entered by copy & paste).
 *
 * NOTE: This routine must be called in "escape mode".
 */
static enum esc_action_t get_action(const struct iucvtty_msg *msg,
				    unsigned char esc)
{
	if (is_esc_char(msg, esc))
		return SEND;

	switch (get_msg_char(msg)) {
	case   0:
		return SEND;
	case '.':
	case 'd':
		return DISCONNECT;
	case 'r':
		return RESIZE;
	default:
		return IGNORE;
	}
}

/**
 * iucvtty_worker() - Handle server connection
 * @terminal:	IUCV TTY server file descriptor
 */
static int iucvtty_worker(int terminal, const struct iucvterm_cfg *cfg)
{
	struct iucvtty_msg *msg;
	fd_set set;
	size_t chunk;
	int in_esc_mode;
	enum esc_action_t action;

	/* setup buffers */
	msg = malloc(MSG_BUFFER_SIZE);
	if (msg == NULL) {
		print_error("Allocating memory for the data buffer failed");
		return -1;
	}

	/* multiplex i/o between login program and socket */
	chunk = 0;
	in_esc_mode = 0;	/* escape mode state */
	action = SEND;
	while (1) {
		if (resize_tty) {
			iucvtty_tx_winsize(terminal, STDIN_FILENO);
			resize_tty = 0;	/* clear signal flag */
		}

		FD_ZERO(&set);
		FD_SET(terminal, &set);
		FD_SET(STDIN_FILENO, &set);

		if (select(MAX(STDIN_FILENO, terminal) + 1, &set,
			   NULL, NULL, NULL) == -1) {
			if (errno == EINTR)
				continue;
			break;
		}

		if (FD_ISSET(terminal, &set)) {
			if (iucvtty_read_msg(terminal, msg,
					     MSG_BUFFER_SIZE, &chunk))
				break;

			switch (msg->type) {
			case MSG_TYPE_DATA:
				iucvtty_copy_data(STDOUT_FILENO, msg);
				write_session_log(msg->data, msg->datalen);
				break;
			case MSG_TYPE_ERROR:
				iucvtty_error(msg);
				break;
			}
		}

		if (FD_ISSET(STDIN_FILENO, &set)) {
			if (iucvtty_read_data(STDIN_FILENO, msg,
					      MSG_BUFFER_SIZE))
				break;

			if (in_esc_mode) {
				in_esc_mode = 0;	/* reset */
				action = get_action(msg, cfg->esc_char);
			} else {
				if (is_esc_char(msg, cfg->esc_char)) {
					in_esc_mode = 1;
					action = IGNORE;
				} else
					action = SEND;
			}

			/* handle escape mode */
			switch (action) {
			case SEND:	/* non-escape mode (default) */
				if (iucvtty_write_msg(terminal, msg))
					goto out_worker_loop;
				break;

			case DISCONNECT:/* disconnect  */
				goto out_worker_loop;

			case RESIZE: 	/* force terminal resize */
				iucvtty_tx_winsize(terminal, STDIN_FILENO);
				break;

			case IGNORE:
				break;
			}
		}
	}

out_worker_loop:
	free(msg);
	return 0;
}

/**
 * main() - IUCV CONN program startup
 */
int main(int argc, char *argv[])
{
	int			rc;
	int 			server;
	struct sockaddr_iucv	addr;
	struct termios		ios;
	struct sigaction	sigact;
	struct passwd		*passwd;
	struct iucvterm_cfg	conf;


	/* gettext initialization */
	gettext_init();

	/* parse command line options */
	parse_options(PROG_IUCV_CONN, &conf, argc, argv);

	/* open session audit log */
	if (conf.sessionlog != NULL)
		if (open_session_log(conf.sessionlog)) {
			print_error("Creating the terminal session "
				    "log files failed");
			return 1;
		}

	/* open socket and connect to server */
	server = iucvtty_socket(&addr, conf.host, conf.service);
	if (server == -1) {
		print_error((errno == EAFNOSUPPORT)
			    ? N_("The AF_IUCV address family is not available")
			    : N_("Creating the AF_IUCV socket failed"));
		return 1;
	}

	/* syslog */
	openlog(SYSLOG_IDENT, LOG_PID, LOG_AUTHPRIV);

	/* get user information for syslog */
	passwd = getpwuid(geteuid());

	if (connect(server, (struct sockaddr *) &addr, sizeof(addr)) == -1) {
		switch (errno) {
		case EAGAIN:
			print_error("The new connection would exceed the "
				    "maximum number of IUCV connections");
			break;
		case ENETUNREACH:
			print_error("The target z/VM guest virtual machine "
				    "is not logged on");
			break;
		case EACCES:
			print_error("The IUCV authorizations do not permit "
				    "connecting to the target z/VM guest");
			break;
		default:
			print_error("Connecting to the z/VM guest virtual "
				    "machine failed");
			break;
		}
		AUDIT("Connection to %s/%s failed for user %s (uid=%i)",
			conf.host, conf.service,
			(passwd != NULL) ? passwd->pw_name : "n/a", geteuid());
		rc = 2;
		goto return_on_error;
	}
	AUDIT("Established connection to %s/%s for user %s (uid=%i)",
		conf.host, conf.service,
		(passwd != NULL) ? passwd->pw_name : "n/a", geteuid());

	/* send client parameters */
	iucvtty_tx_termenv(server, DEFAULT_TERM);
	iucvtty_tx_winsize(server, STDIN_FILENO);

	/* register signal handler */
	sigemptyset(&sigact.sa_mask);
	sigact.sa_flags = SA_RESTART;
	sigact.sa_handler = sig_handler;
	sigaction(SIGWINCH, &sigact, NULL);
	sigaction(SIGTERM,  &sigact, NULL);

	/* modify terminal settings */
	if (tcgetattr(STDIN_FILENO, &ios_orig)) {
		print_error("Getting the terminal I/O settings failed");
		rc = 3;
		goto return_on_error;
	}
	memcpy(&ios, &ios_orig, sizeof(ios));

	/* put terminal into raw mode */
	cfmakeraw(&ios);
	/* NOTE: If the TTY driver (ldisc) runs in TTY_DRIVER_REAL_RAW,
	 *       we need to do the input character processing here;
	 *       that means to translate CR into CR + NL (ICRNL).
	 * Define TTY_REAL_RAW in for that case. */
#ifdef TTY_REAL_RAW
	ios.c_iflag |= ICRNL;			/* | IGNPAR | IGNBRK; */
#endif
	tcflush(STDIN_FILENO, TCIOFLUSH);
	if (tcsetattr(STDIN_FILENO, TCSANOW, &ios)) {
		print_error("Modifying the terminal I/O settings failed");
		rc = 4;
		goto return_on_error;
	}

	iucvtty_worker(server, &conf);

	tcsetattr(STDIN_FILENO, TCSANOW, &ios_orig);

	rc = 0;
return_on_error:
	close(server);
	closelog();
	close_session_log();

	return rc;
}
